examples/epocher/src/application/actor.rs

```diff
@@ -0,0 +1,260 @@
+                            EpochCert::Single(f0)
+                        }
+                        _ => {
+                            let Some(f2) = self.get_finalization(height).await else {
+                                continue;
+                            };
+                            let previous_height = epoch::get_last_height(epoch - 1);
+                            let Some(f1) = self.get_finalization(previous_height).await else {
+                                continue;
+                            };
+                            EpochCert::Double(f1, f2)
+                        }
+                    };
+
+                    orchestrator.report(cert).await;
+
+                    let _ = response.send(());
+                }
+            }
+        }
+    }
+
+    /// Gets the finalization for a given height from marshal.
+    async fn get_finalization(
+        &mut self,
+        height: u64,
+    ) -> Option<Finalization<MinSig, Sha256Digest>> {
+        let Ok(Some((finalization, _block))) = self.marshal.get_finalization(height).await.await
+        else {
+            warn!("finalization not found for height {}", height);
+            return None;
+        };
+        Some(finalization)
+    }
+
+    /// Proposes a new block.
+    async fn propose_block(&mut self, context: Context<Sha256Digest>) -> Sha256Digest {
+        let round = context.round;
+        let epoch = round.epoch();
+        let (parent_view, parent_commitment) = context.parent;
+        let parent_round = Round::new(epoch, parent_view);
+
+        // Get the parent block from marshal.
+        let parent = self.subscribe_block(parent_round, parent_commitment).await;
+
+        // Case: Reproposal.
+        if parent.height() == epoch::get_last_height(epoch) {
+            debug!(payload=?parent_commitment, "propose: repropose");
+            return parent_commitment;
+        }
+
+        // Case: New proposal.
+        let height = parent.height() + 1;
+        let random = self.context.gen::<u64>();
+        let block = Block::new(parent_commitment, height, random);
+        let digest = block.commitment();
+        self.marshal.broadcast(block.clone()).await;
+        self.marshal.verified(round, block).await;
+        debug!(payload=?digest, "propose: new-proposal");
+        digest
+    }
+
+    /// Verifies a block.
+    async fn verify_block(
+        &mut self,
+        context: Context<Sha256Digest>,
+        payload: Sha256Digest,
+    ) -> bool {
+        // Get the proposed block from marshal.
+        let round = context.round;
+        let epoch = round.epoch();
+        let block = self.subscribe_block(round, payload).await;
+        let height = block.height();
+
+        // Ensure that the height is appropriate for the epoch.
+        if !epoch::height_in_epoch(height, epoch) {
+            debug!(payload=?payload, "verify: rejected-height-not-in-epoch");
+            return false;
+        }
+
+        // Case: Reproposal.
```

Comment:
Like my comment around propose, I wonder if we should try to hide this (as nothing good can happen).